forte/data/data_pack_dataset.py

```diff
@@ -0,0 +1,278 @@
+from forte.data.converter import Feature
+from forte.data.data_pack import DataPack
+from forte.data.extractor.base_extractor import BaseExtractor
+from forte.data.ontology.core import EntryType
+from forte.data.ontology.top import Annotation
+from forte.data.types import DataRequest
+
+__all__ = [
+    "DataPackIterator",
+    "DataPackDataSource",
+    "DataPackDataset"
+]
+
+# An instance is a single data point from data pack
+Instance = Dict
+RawExample = Tuple[Instance, DataPack]
+FeatureCollection = Dict[str, Feature]
+
+
+class DataPackIterator:
+    """
+    An iterator over single data example from multiple data packs.
+
+    Args:
+        pack_generator (Iterator[DataPack]): A generator of
+            :class:`forte.data.data_pack.DataPack`.
+        context_type: The granularity of a single example which
+            could be any ``Annotation`` type. For example, it can be
+            :class:`ft.onto.base_ontology.Sentence`, then each training example
+            will represent the information of a sentence.
+        request: The request of type `Dict` sent to
+            :class:`forte.data.readers.base_reader.PackReader` to query
+            specific data.
+        skip_k (int): Will skip the first `skip_k` instances and generate
+            data from the (`offset` + 1)th instance.
+
+    .. note::
+        For parameters `context_type`, `request`, `skip_k`, please refer to
+        :meth:`get_data()` in :class:`forte.data.data_pack.DataPack`.
+    """
+    def __init__(self,
+                 pack_generator: Iterator[DataPack],
+                 context_type: Type[Annotation],
+                 request: Optional[DataRequest] = None,
+                 skip_k: int = 0):
+        self._get_data_args: Dict = {
+            "context_type": context_type,
+            "request": request,
+            "skip_k": skip_k
+        }
+
+        self._data_pack_iter: Iterator[DataPack] = pack_generator
+        self._instance_iter: Optional[Iterator[Instance]] = None
+        self._curr_data_pack: Optional[DataPack] = None
+
+    def __iter__(self):
+        return self
+
+    def __next__(self) -> RawExample:
+        if self._curr_data_pack is None:
+            self._curr_data_pack = next(self._data_pack_iter)
+            self._instance_iter = \
+                self._curr_data_pack.get_data(**self._get_data_args)
+
+        assert self._instance_iter is not None
+
+        try:
+            return next(self._instance_iter), self._curr_data_pack
+        except StopIteration:
+            # Current data pack has no more instance. Go to next data pack.
+            self._curr_data_pack = next(self._data_pack_iter)
+            self._instance_iter = \
+                self._curr_data_pack.get_data(**self._get_data_args)
+
+        return next(self._instance_iter), self._curr_data_pack
+
+
+class DataPackDataSource(IterDataSource):
+    """
+    A data source consists of data packs. It contains an iterator over
```

Comment:
You should probably say:

A texar `IterDataSource` that returns data from data packs. Internally, it wraps a :class:`
~forte.data.data_pack_dataset.DataPackIterator`

"consists of" does not sound very accurate here.

Two more notes on documentation:
- Adding `~` before the class name will make sphinx only renders the short name, which is most of time desired.
- It is better to link to Texar doc here, because we have specified the inter sphinx link (https://github.com/asyml/forte/blob/master/docs/conf.py#L324)

Reply:
Fixed. All module references are shorten in this PR and all texar module references are linked via sphinx.